src/server/src/main/java/io/cassandrareaper/jmx/JmxConnectionFactory.java

```diff
@@ -68,9 +70,36 @@ protected JmxProxy connect(Optional<RepairStatusHandler> handler, String host, i
       password = jmxAuth.getPassword();
     }
     try {
-      JmxProxy jmxProxy = JmxProxyImpl.connect(handler, host, username, password, addressTranslator, connectionTimeout);
-      hostConnectionCounters.incrementSuccessfulConnections(host);
-      return jmxProxy;
+      if (JMX_CONNECTIONS.containsKey(host) && JMX_CONNECTIONS.get(host).isConnectionAlive()) {
+        return JMX_CONNECTIONS.get(host);
+      }
+
+      JmxProxy previousConnection = JMX_CONNECTIONS.getOrDefault(host, null);
+
+      if (previousConnection == null
+          || (previousConnection != null && !previousConnection.isConnectionAlive())) {
+        // There is no connection in the map for that host or it died already
+        synchronized (this) {
+          // Double checking now that we're in the sync block
+          JmxProxy prevConnection = JMX_CONNECTIONS.getOrDefault(host, null);
+          if (prevConnection != null && !previousConnection.isConnectionAlive()) {
+            // connection exists but is not live anymore
+            previousConnection.close();
+            JMX_CONNECTIONS.remove(host);
+          }
+          if (prevConnection == null) {
+            LOG.info("Adding new JMX Proxy for host {}", host);
+
+            JMX_CONNECTIONS.putIfAbsent(
+                host,
+                JmxProxyImpl.connect(
+                    host, username, password, addressTranslator, connectionTimeout));
+            hostConnectionCounters.incrementSuccessfulConnections(host);
+          }
+        }
+      }
+
+      return JMX_CONNECTIONS.get(host);
```

Comment:
the whole `connect(..)` method could be written a lot simpler with:

```
  protected JmxProxy connect(Optional<RepairStatusHandler> handler, String host, int connectionTimeout)
      throws ReaperException, InterruptedException {
    // use configured jmx port for host if provided
    if (jmxPorts != null && jmxPorts.containsKey(host) && !host.contains(":")) {
      host = host + ":" + jmxPorts.get(host);
    }

    String username = (jmxAuth != null) ? jmxAuth.getUsername() : null;
    String password = (jmxAuth != null) ? jmxAuth.getPassword() : null;

    try {
      JmxConnectionProvider provider = new JmxConnectionProvider(host, username, password, connectionTimeout);
      JMX_CONNECTIONS.computeIfAbsent(host, provider);
      JmxProxy proxy = JMX_CONNECTIONS.get(host);
      if (!proxy.isConnectionAlive()) {
        LOG.info("Adding new JMX Proxy for host {}", host);
        JMX_CONNECTIONS.put(host, provider.apply(host));
        proxy.close();
      }
      return JMX_CONNECTIONS.get(host);
    } catch (RuntimeException ex) {
      // unpack any exception behind JmxConnectionProvider.apply(..)
      if (ex.getCause() instanceof InterruptedException) {
        throw (InterruptedException) ex.getCause();
      }
      if (ex.getCause() instanceof ReaperException) {
        throw (ReaperException) ex.getCause();
      }
      throw ex;
    }
  }
```

along with adding the inner class:
```
  private class JmxConnectionProvider implements Function<String, JmxProxy> {

    private final String host;
    private final String username;
    private final String password;
    private final int connectionTimeout;

    public JmxConnectionProvider(String host, String username, String password, int connectionTimeout) {
      this.host = host;
      this.username = username;
      this.password = password;
      this.connectionTimeout = connectionTimeout;
    }

    @Override
    public JmxProxy apply(String host) {
      Preconditions.checkArgument(host.equals(this.host));
      try {
        JmxProxy proxy = JmxProxyImpl.connect(host, username, password, addressTranslator, connectionTimeout);
        hostConnectionCounters.incrementSuccessfulConnections(host);
        return proxy;
      } catch (ReaperException | InterruptedException ex) {
        hostConnectionCounters.decrementSuccessfulConnections(host);
        throw new RuntimeException(ex);
      }
    }
  }
```